### PR TITLE
Occ command to update group mapping information

### PR DESCRIPTION
**Pull request by jvillafanez**
Original PR: https://github.com/owncloud/core/pull/26029

Related issue: https://github.com/owncloud/enterprise/issues/1464

We need to agree on backporting this PR (issue comes from version 8.2)

### CODE REVIEW

File: lib/Command/UpdateGroup.php [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+			$usersAdded = array_diff($userList, $mappedList);
+			$usersRemoved = array_diff($mappedList, $userList);
+			return array('added' => $usersAdded, 'removed' => $usersRemoved);
+		}
+	}
+
+	/**
+	 * Make sure $groupNames doesn't contain duplicated values. This function could behave
+	 * unexpectedly otherwise.
+
```
Comment (PVince81) [inline lib/Command/UpdateGroup.php]: the invasion of the red blood spaces has begun!

**Comment by PVince81** [comment]
- [ ] BUG: Display a message "no changes" when no changes are found for any of the selected groups ? (I ran it and was confused to see no output at all)

**Comment by PVince81** [comment]
Hmmm, no sure what I'm doing wrong: whatever the situation there is never any output.

I have the following config:
- openLDAP (the dockers from https://github.com/owncloud/administration/tree/master/ldap-testing with the zombie army)
- I've added a few zombies in the groups "Box10" and "Box100" using the memberUid attribute
- LDAP config has memberUid configured

Then I run this:
```
occ ldap:update-group Box10 -vvvvvvv
```
No output.
Then remove or add some users in the group Box10 in LDAP directly.
Then run the command again: no output either.

Only if I pass a wrong group id I'll get an error, so output does work.

I'm unable to paste my LDAP config as show-config is broken and I don't know where to find the config-id in the DB: https://github.com/owncloud/user_ldap/issues/16

**Comment by PVince81** [comment]
Hmm okay, seems I passed to many `v` to the verbosity. Would be good if it could still work as other commands like occ files:scan accept any number of "v".

Here is the output after adding and deleting a user from "Box10":
```
group membership attribute is critical for this command to work properly, please verify
* localhost:389 -> memberUid
checking group "Box10"...
updating "Box10" group DB information
sending hooks
new users:
8c5e411a-1eba-1036-991b-57e71cf45853
removed users:
8cc327d8-1eba-1036-99a5-57e71cf45853
```

Hmm, ok so it seems to work.

- [ ] BUG: display at least a summary output when no verbosity given, else one believes that it didn't work

File: lib/Command/UpdateGroup.php [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	protected function execute(InputInterface $input, OutputInterface $output) {
+		$groupIDs = $input->getArgument('groupID');
+		// make sure we don't have duplicated groups in the parameters
+		$groupIDs = array_unique($groupIDs);
+
+		$helper = $this->helper;
+		$availableConfigs = $helper->getServerConfigurationPrefixes();
+
+		// show configuration information, useful to debug
+		$output->writeln('group membership attribute is critical for this command to work properly, please verify', OutputInterface::VERBOSITY_VERBOSE);
```
Comment (jvillafanez) [inline lib/Command/UpdateGroup.php]: I didn't want to keep bothering the user over and over when the attribute is properly set. However, adding the `--yes` option seems a better choice.

File: lib/Command/UpdateGroup.php [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+		$this->connection = $connection;
+		$this->ldap = $ldap;
+		$this->helper = $helper;
+		parent::__construct();
+	}
+
+	protected function configure() {
+		$this
+			->setName('ldap:update-group')
+			->setDescription('update the specified group information')
```
Comment (PVince81) [inline lib/Command/UpdateGroup.php]: sounds good

File: lib/Command/UpdateGroup.php (hunk 2 above)
Comment (PVince81) [inline lib/Command/UpdateGroup.php]: this, assuming that we cannot detect whether the property is properly set.

Comment (PVince81) [inline lib/Command/UpdateGroup.php]: now thinking of it I think it's ok to keep the warning but not ask for confirmation.

If an admin really goes as far as using this command, it means they already have setup their groups in LDAP properly so there is no need to bother that much.

**Comment by jvillafanez** [comment]
Adjusted messages, plus the verbosity of some of them in order to show output when no verbosity is given.

There is one thing that might not be a good idea: the confirmation dialog is also hidden if the command runs with the "-q" option. The command hangs until the user confirms the hidden dialog, which is a very bad UX from my point of view.
I don't think there is a good solution for this other than blame the user's stupidity or just get rid of the confirmation dialog and just run the command.

If we decide to get rid of the dialog, we might need to show the warning message always (previous verbosity to make it appear was "-v") and add another option to hide the warning.

**Comment by PVince81** [comment]
Yeah, please just get rid of the confirmation. I think it's not needed as per my comment here https://github.com/owncloud/user_ldap/pull/14#discussion_r81910180

If an admin made it that far, then they already have proper group mappings.

**Comment by PVince81** [comment]
@jvillafanez let me know when this is ready for review again and set the label to "3 - To Review". Thanks

**Comment by PVince81** [comment]
Verbosity is better now.
But when there is no change it looks like this:
```
checking group "Box10"...
new users:
removed users:
checking group "Box100"...
new users:
removed users:
```

Also I hope no one will get the idea trying to parse this output.
I'm fine keeping it this way for now.

**Comment by PVince81** [comment]
- [ ] BUG: group deletion :bomb: 

I deleted "Box100" which existed in openLDAP, then ran the command:
```
"Box100" doesn't exist in LDAP any more, removing local mapping

                                                                                                                                                 
  [Doctrine\DBAL\Exception\SyntaxErrorException]                                                                                                 
  An exception occurred while executing 'DELETE FROM `oc_ldap_group_mapping` lgm WHERE lgm.`owncloud_name` = ?' with params ["Box100"]:          
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your Maria  
  DB server version for the right syntax to use near 'lgm WHERE lgm.`owncloud_name` = 'Box100'' at line 1                                        
                                                                                                                                                 

                                                                                                                                                 
  [Doctrine\DBAL\Driver\PDOException]                                                                                                            
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your Maria  
  DB server version for the right syntax to use near 'lgm WHERE lgm.`owncloud_name` = 'Box100'' at line 1                                        
                                                                                                                                                 

                                                                                                                                                 
  [PDOException]                                                                                                                                 
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your Maria  
  DB server version for the right syntax to use near 'lgm WHERE lgm.`owncloud_name` = 'Box100'' at line 1                                        
```

**Comment by jvillafanez** [comment]
> Also I hope no one will get the idea trying to parse this output.

I tried to keep the changes as minimum as possible, so showing a trace instead of a parseable output was easier. If it's needed then we should fix it in another PR.

> BUG: group deletion

Solved now.

**Comment by PVince81** [comment]
Tested, works :+1: 

Question: should this command be extended to also allow adding/syncing newly created groups that weren't known in OC before ? (or ones that were deleted and re-added)

**Comment by jvillafanez** [comment]
> Question: should this command be extended to also allow adding/syncing newly created groups that weren't known in OC before ? (or ones that were deleted and re-added)

Probably, but taking into account we'll have to perform a search which might bring more groups (searching for Box1 will bring also Box11, Box12, Box100, etc), and I'm not sure about possible side effects specially with caching, I don't know if it's a good idea.

**Comment by DeepDiver1975** [comment]
@jvillafanez please press the button for the CLA - [CLA assistant check](https://cla-assistant.io/owncloud/user_ldap?pullRequest=14)

**Comment by PVince81** [comment]
looks like he did now, let's merge

**Comment by PVince81** [comment]
As this is a new feature, I'm not sure about backporting... is this critical enough ? @felixboehm

**Comment by felixboehm** [comment]
Yes definitiv for 9.0

**Comment by DeepDiver1975** [comment]
> Yes definitiv for 9.0

@jvillafanez can i ask you to take care? thx

**Comment by jvillafanez** [comment]
Sure, taking into account the set of changes, I'll just copy & paste the file between the repos (with the minor change in the register_command.php file)

**Comment by jvillafanez** [comment]
Stable9.1: https://github.com/owncloud/core/pull/26341
Stable9: https://github.com/owncloud/core/pull/26343

**Comment by DeepDiver1975** [comment]
> Stable9.1: owncloud/core#26341
> Stable9: owncloud/core#26343

@felixboehm

**Comment by felixboehm** [comment]
Great work, thanks!

**Comment by PVince81** [comment]
Raised https://github.com/owncloud/user_ldap/issues/20 because I think that we shouldn't fail in case one group is not found, due to potential admin + cron use cases.

**Pull request merged**